### DISCUSSION TITLE
setMediaElement not updating progress bar

### DISCUSSION THREAD

**Discussion by jahangiranwari** [katspaugh/wavesurfer.js#3266] [General]
With the new [setMediaElement](https://wavesurfer.xyz/docs/classes/wavesurfer.default#setMediaElement) feature to attach a `<audio>` at runtime we noticed that although we are able to share the same underlying HTMLMediaElement element between two players it doesn't update the progress bar on both players.

To demonstrate the issue and using [React Global Player](https://wavesurfer.xyz/examples/?react-global-player.js) example as reference created this [gist](https://gist.github.com/jahangiranwari/e461e6f8058dfaffb1092f72c7a88282). When we click Play the track starts playing in the global player and we can seek on either player indicating that they are sharing the same underlying `<audio>` element. But the track player's progress bar doesn't move.

On deeper investigation identified the root cause as missing `timeupdate` listener on the newly attached HTMLMediaElement. The same can be confirmed by uncommenting [these lines](https://gist.github.com/jahangiranwari/e461e6f8058dfaffb1092f72c7a88282#file-wavesurfer-jsx-L146-L149) in the same gist.

### Possible Solutions
One possible solution could be to track the `timeupdate` listener independently of [other subscriptions](https://github.com/katspaugh/wavesurfer.js/blob/8eba6965f13bcaa35b10d024c7064d4e1602e26c/src/wavesurfer.ts#L133) and subscribe/unsubscribe to it in `setMediaElement`. 

Another solution could be to have public methods that would allow us to subscribe/unsubscribe from all player events currently used in [initPlayerEvents](https://github.com/katspaugh/wavesurfer.js/blob/8eba6965f13bcaa35b10d024c7064d4e1602e26c/src/wavesurfer.ts#L179).

**Comment by katspaugh** [comment]
Good catch. I think adding setMediaElement was a bad idea. It essentially requires a re-initialization of the entire waveform and all the events, so I don’t see the point of such a setter.

As you can see in the original global player example, it’s possible to simply re-create a wavesurfer instance.

**Reply by jahangiranwari** [reply]
In a complex application where many (e.g. 15+) wavesurfer instances are rendered using per-generated peaks data and we have two players (Global & PlaylistTrack) that need to be in sync recreating wavesurfer instance can get expensive, especially if a user is exploring a large music catalog performing various actions on different tracks like play/pause/seek. 

With `setMediaElement` we avoid the costs of recreating _(and re-rendering)_ the whole wavesurfer instance. The unfortunate blocker is that the progress bar doesn't work. In experiments we performed so far using `setMediaElement`  doesn't require reinitalizing the entire waveform rather only the "player events".

**Reply by katspaugh** [reply]
You have to re-render it anyway, because if you switch to another track, it's a different waveform. You can pass the pre-decoded peaks from the track in your playlist to the global player to avoid fetching and decoding the audio again.

**Reply by jahangiranwari** [reply]
Yes, we do need to re-render the Global player. However, the playlist track player doesn't need to be re-rendered.

Using your example from a previous discussion:
<img src="https://github.com/katspaugh/wavesurfer.js/assets/1464320/47eae86c-df83-4c4f-a384-7f45e9d67f30" width=70% height=70%>

The flow of events would be something like:

1. User lands on Homepage
2. Multiple (e.g. 10+) track players are rendered using pre-generated waveform
3. User clicks on player 1
    - The track  1 audio URL and pre-generated peaks data is passed to global player
4. On player 1 the `media` is set to global player's media
    - now global player and player 1 are sharing the same `HTMLAudioElement`
5. User performs play/pause/seek on either global player or player 1 and both player remain in sync
6. User decides to hear another track and clicks on player 3
   - The track  3 audio URL and pre-generated peaks data is passed to global player
8. On player 1 the `media` is detached from global player's media and attached back to its original `audio`
    - player 1 is now back to its original state (as it was on homepage load) without requiring any re-renders
9. On player 3 the `media` is set to global player's media
    - now global player and player 3 are sharing the same `HTMLAudioElement`
10. User performs play/pause/seek on either global player or player 3 and both player remain in sync

This way we only have to re-render the global player on track changes and avoid having to recreate/re-render the playlist track players which significantly improves the overall performance of the app.

**Reply by katspaugh** [reply]
Oh, I see. You're setting the media on the playlist tracks, not on the global player.
If you reverse this flow, you will only need to re-render the global player.
I.e. don't change the playlist tracks but re-create the global player with the media element, URL and peaks from the currently playing playlist track.

**Reply by jahangiranwari** [reply]
That's correct. We initially went with that approach. However, that approach breaks down when we want to persist the global player across screens navigation. If the global player is referencing a track player `media`, once the user navigates to a different screen the track player's `media` is destroyed and global player stops playing.  This was the reason we had to use `setMediaElement` on playlist track player.

**Reply by katspaugh** [reply]
You can avoid destroying it if you pass an external media to your playlist tracks. Wavesurfer in that case won't touch it on destroy.
E.g.

```js
const audio = new Audio()
const playlistTrack = WaveSurfer.create({
  media: audio,
  // ...
})
```

**Reply by jahangiranwari** [reply]
We experimented with that as well and realized that it resulted in a lot of zombie `audio` objects lying around as the user navigates through different screens and in some cases when the user navigated away and came back to the same screen we were creating duplicate `audio` objects for the same playlist track player.

**Reply by katspaugh** [reply]
If you just create an audo w/o adding it to the DOM, it should be garbage-collected as soon as the last reference to it is deleted. As long as you're destroying your wavesurfer instances correctly, only the global player will keep a reference to the currently played audio.

**Reply by jahangiranwari** [reply]
Thanks for the suggestion. We experimented changing the flow of setting the `media` on global player instead of track player. Although it works for on-page track changes and screen navigation it breaks our "Playlist Play" feature where the user is able to continue listening to playlist even after they navigate away from playlist screen. This is because once user navigates away from playlist page the previous/next track's `audio` is no longer available as it has been garbage collected. 

We observed another side-effect of using external media, between screen navigation the global player stops playing. On deeper investigation we discovered that when `destroy` of wavesurfer instance is called we [pause](https://github.com/katspaugh/wavesurfer.js/blob/8eba6965f13bcaa35b10d024c7064d4e1602e26c/src/player.ts#L76) the current audio.

**Reply by katspaugh** [reply]
I should probably fix this, it shouldn’t pause it.

**Comment by jahangiranwari** [comment]
@katspaugh is it safe to assume that `setMediaElement` would be retained? Even with its limitation it appears to be best suited for complex apps where performance is critical and we would be relying on it heavily in our app.

**Reply by katspaugh** [reply]
It needs to be either fixed (that is, events subscribed and unsubscribed properly when a new media is set), or removed. Adding it in the current state was a mistake.

**Reply by jahangiranwari** [reply]
Thoughts on fixing it as shown below?

```javascript
class WaveSurfer extends Player<WaveSurferEvents> {
  protected mediSubscriptions: Array<() => void> = []

  // ...
  // ...

  private initPlayerEvents() {
    this.mediSubscriptions.push(
      this.onMediaEvent("timeupdate", () => {
        const currentTime = this.getCurrentTime()
        this.renderer.renderProgress(currentTime / this.getDuration(), this.isPlaying())
        this.emit("timeupdate", currentTime)
      }),

      this.onMediaEvent("play", () => {
        this.emit("play")
        this.timer.start()
      }),

      this.onMediaEvent("pause", () => {
        this.emit("pause")
        this.timer.stop()
      }),

      this.onMediaEvent("emptied", () => {
        this.timer.stop()
      }),

      this.onMediaEvent("ended", () => {
        this.emit("finish")
      }),

      this.onMediaEvent("seeking", () => {
        this.emit("seeking", this.getCurrentTime())
      })
    )
  }

  private unsubscribePlayerEvents() {
    this.mediSubscriptions.forEach((unsubscribe) => unsubscribe())
  }

  public setMediaElement(element: HTMLMediaElement) {
    this.unsubscribePlayerEvents()
    super.setMediaElement(element)
    this.initPlayerEvents()
  }

  public destroy() {
    this.emit("destroy")
    this.plugins.forEach((plugin) => plugin.destroy())
    this.subscriptions.forEach((unsubscribe) => unsubscribe())
    this.unsubscribePlayerEvents()
    this.timer.destroy()
    this.renderer.destroy()
    super.destroy()
  }
}
```

**Reply by katspaugh** [reply]
Yeah, looks good to me!
The only thing I would change is to clear the `this.mediSubscriptions` array after unsubscribing.

**Reply by jahangiranwari** [reply]
Sound good. Will work on a PR.

**Reply by katspaugh** [reply]
Appreciate the PR! I’ve merged it and released 7.3.9. Kudos!

**Reply by jahangiranwari** [reply]
Thank you for the quick release.
